guides/05_chatbots/03_agents-and-tool-usage.md

```diff
@@ -166,4 +167,139 @@ demo.launch()
+
+
+### A real example using Gemini2.0 Flash Thinking API
+
+Let's create a complete chatbot that shows its thoughts and responses in real-time. We'll use Google's Gemini API for accessing Gemini2.0 Flash Thinking LLM and Gradio for the UI.
+
+We'll begin with imports and setting up the gemini client. Note that you'll need to acquire a Google Gemini API key first -
+
+```python
+import gradio as gr
+from gradio import ChatMessage
+from typing import Iterator
+import google.generativeai as genai
+
+genai.configure(api_key="your-gemini-api-key")
+model = genai.GenerativeModel("gemini-2.0-flash-thinking-exp-1219")
+```
+
+First, let's set up our streaming function that handles the model's output:
+
+```python
+def stream_gemini_response(user_message: str, messages: list) -> Iterator[list]:
+    """
+    Streams both thoughts and responses from the Gemini model.
+    """
+    # Initialize response from Gemini
+    response = model.generate_content(user_message, stream=True)
+    
+    # Initialize buffers
+    thought_buffer = ""
+    response_buffer = ""
+    thinking_complete = False
+    
+    # Add initial thinking message
+    messages.append(
+        ChatMessage(
+            role="assistant",
+            content="",
+            metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
+        )
+    )
+    
+    for chunk in response:
+        parts = chunk.candidates[0].content.parts
+        current_chunk = parts[0].text
+        
+        if len(parts) == 2 and not thinking_complete:
+            # Complete thought and start response
+            thought_buffer += current_chunk
+            messages[-1] = ChatMessage(
+                role="assistant",
+                content=thought_buffer,
+                metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
+            )
+            
+            # Add response message
+            messages.append(
+                ChatMessage(
+                    role="assistant",
+                    content=parts[1].text
+                )
+            )
+            thinking_complete = True
+            
+        elif thinking_complete:
+            # Continue streaming response
+            response_buffer += current_chunk
+            messages[-1] = ChatMessage(
+                role="assistant",
+                content=response_buffer
+            )
+            
+        else:
+            # Continue streaming thoughts
+            thought_buffer += current_chunk
+            messages[-1] = ChatMessage(
+                role="assistant",
+                content=thought_buffer,
+                metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
+            )
+        
+        yield messages
+```
+
+Then, let's create the Gradio interface:
```

Comment:
Could we use `gr.ChatInterface` instead? `gr.ChatInterface` now supports the function returning multiple messages (you just return or yield a `list`), which would simplify the Gradio code and provide a better UI.

Reply:
I was thinking that a more granular implementation with `gr.Chatbot` would add more value to the guide. But there are no strong opinions here.